docs/index.rst

```diff
@@ -1,260 +1,427 @@
+   will serve as our default.
+3. Read a configuration from a `YAML`_ file in the system configuration
+   directory.
+4. Compose the two configurations by laying the system configuration over the
+   default configuration. Every variable in the system configuration will
+   override the ones in the default.
+5. Read a string variable from the configuration.
+
+.. _YAML: http://yaml.org/
+
+This example is just enough to get a taste for Confit. Details are explained
+in later sections.
+
+
+Git
+^^^
+
+To demonstrate the full power of Confit, consider the `configuration model
+for Git <http://git-scm.com/docs/git-config#FILES>`_.
+
+::
+
+  from_git_ini = lambda path: confit.from_ini(path, sep=".")
+  sys_config = from_git_ini(confit.sys_config_path("git", "config"))
+  second_user_config = sum(map(from_git_ini, confit.user_config_paths("git", "config")))
+  first_user_config = from_git_ini(os.expanduser("~/.gitconfig"))
+  repo_config = from_git_ini(os.path.join(git_dir(), "config"))
+  # Git takes configuration from environment variables with names different
+  # from the configuration variables they map to. That requires some custom
+  # (but simple) logic not built in to Confit. It is left as an exercise for
+  # the reader.
+  env_config = ...
+  config = sum((defaults, sys_config, second_user_config, first_user_config, repo_config, env_config))
+
+
+curl
+^^^^
+
+Confit is useful for libraries in addition to applications. Consider libcurl,
+which loads configuration from `~/.curlrc` and special environment variables
+(`http_proxy`, `https_proxy`, `no_proxy`, etc). Libraries can use all the
+facilities of Confit (with the exception of ``from_args``) to ease handling
+of their configuration.
+
+
+Configuration
+-------------
+
+A configuration consists of variable bindings. In each new project, a
+configuration is born when some variable holds an arbitrary value, e.g.  a
+path to a temporary directory. The value is moved to the configuration as a
+default, and the variable's value is loaded from the configuration.
+
+Variables
+^^^^^^^^^
+
+In Confit, a configuration is an immutable mapping from string names to
+*variables*. Variables have two methods, ``get`` and ``origin``.
+
+::
+
+  get(cast=lambda x: x)
+
+Call *cast* with the variable's value and return the result. *cast* is useful
+for casting the value to a different type (i.e., "get the value as a ___").
+*cast* may also check the value against some validity criteria, e.g. whether
+a string can be parsed, or whether an integer fits within a range. If *cast*
+raises an exception, then it will be wrapped in a ``ConfitCastError``, that,
+if uncaught, will be pretty-printed to the console (using ``sys.excepthook``)
+indicating both the failing call to ``get`` and the origin of the faulty
+value::
+
+  /usr/bin/script:123:
+      port = config['port'].get(is_not_privileged)
+  failed on the override at /home/user/.example.yml:8:
+  port: 80
```

Comment:
I'm cool with that. Maybe we can include a function to register a standard pretty printing exception hook for the convenience of users who want it?